### PR TITLE
[Enhancement] Support adjust cachesize automatically

### PR DESCRIPTION
**Pull request by zaorangyang**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

The BE memory is used by many logics, such as queries, ingestions, compactions, pagecache, and so on. pagecache is enabled by default, and the user can adjust the maximum amount of memory. Furthermore, we would like to have a pagecache self-adjustment policy, so that when memory pressure is low, pagecache can buffer more data, and when memory pressure is high, Pagecache can buffer less data. pagecache can actively evict some data to ensure the running of BE.

This PR automatically resize the Pagecache by monitoring the current memory water level. Below is a test result with auto-tuning cachesize enabled. This test specifies a urgent water level of 30% and a high water level of 20%. The memory usage fluctuates between urgent and high water levels. 
image

If auto-tuning cachesize is turned off, memory usage is not constrained.
image

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: be/src/common/config.h [hunk 1]
```diff
@@ -52,6 +52,18 @@ CONF_mInt64(tc_free_memory_rate, "0");
 // tcmalloc gc period, default 60, it should be between [1, 180]
 CONF_mInt64(tc_gc_period, "60");
 
+CONF_mBool(enable_auto_adjust_pagecache, "true");
+// Memory urget water level, if the memory usage exceeds this level, reduce the size of
+// the Pagecache immediately, it should be between (memory_high_level, 100].
```
Comment (trueeyu) [inline be/src/common/config.h:57]: it should between

File: be/src/common/config.h [hunk 2]
```diff
@@ -52,6 +52,18 @@ CONF_mInt64(tc_free_memory_rate, "0");
...
+CONF_mInt64(memory_urgent_level, "85");
+// Memory high water level, if the memory usage exceeds this level, reduce the size of
+// the Pagecache slowly, it should be between [1, memory_urgent_level).
+CONF_mInt64(memory_high_level, "75");
+// Pagecache size adjust period, default 20, it should be between [1, 180].
+CONF_Int64(pagecache_adjuct_period, "20");
+// Sleep time in seconds between pagecache adjust iterations.
+CONF_Int64(auto_adjust_pagecache_interval, "10");
```
Comment (trueeyu) [inline be/src/common/config.h:65]: _seconds

File: be/src/util/lru_cache.cpp [hunk 3]
```diff
@@ -402,16 +402,30 @@ ShardedLRUCache::ShardedLRUCache(size_t capacity) : _last_id(0), _capacity(capac
...
 }
 
+void ShardedLRUCache::set_capacity(size_t capacity) {
+    // Maybe multi client try to set capactity, we protect it using mutex.
+    std::lock_guard l(_mutex);
+    _set_capacity(capacity);
+}
+
+bool ShardedLRUCache::adjust_capacity(int64_t delta, size_t min_capacity) {
+    std::lock_guard l(_mutex);
```
Comment (trueeyu) [inline be/src/util/lru_cache.cpp:420]: If large memory is released at one time, will the lock time here be longer?

File: be/src/storage/olap_server.cpp [hunk 4]
```diff
@@ -176,10 +177,91 @@ Status StorageEngine::start_bg_threads() {
...
+#endif
+    GCHelper dec_advisor(config::pagecache_adjuct_period, config::auto_adjust_pagecache_interval, MonoTime::Now());
+    GCHelper inc_advisor(config::pagecache_adjuct_period, config::auto_adjust_pagecache_interval, MonoTime::Now());
+    auto cache = StoragePageCache::instance();
+    while (!_bg_worker_stopped.load(std::memory_order_consume)) {
+        SLEEP_IN_BG_WORKER(config::auto_adjust_pagecache_interval);
+        if (!config::enable_auto_adjust_pagecache) {
+            continue;
+        }
+        if (config::disable_storage_page_cache) {
```
Comment (trueeyu) [inline be/src/storage/olap_server.cpp:216]: No need to start the gc thread, if `disable_storage_page_cache`=true?

File: be/src/storage/olap_server.cpp [hunk 5]
```diff
@@ -176,10 +177,91 @@ Status StorageEngine::start_bg_threads() {
...
+        }
+        MemTracker* memtracker = ExecEnv::GetInstance()->process_mem_tracker();
+        if (memtracker == nullptr || !memtracker->has_limit() || cache == nullptr) {
+            continue;
+        }
+
+        // Check config valid
+        int64_t memory_urgent_level = config::memory_urgent_level;
+        int64_t memory_high_level = config::memory_high_level;
+        if (UNLIKELY(!(memory_urgent_level > memory_high_level && memory_high_level >= 1 &&
```
Comment (trueeyu) [inline be/src/storage/olap_server.cpp:227]: Check the config at be start or manual modification?

**Review by trueeyu — DISMISSED** [review]

File: be/src/common/config.h [hunk 6]
```diff
@@ -52,6 +52,18 @@ CONF_mInt64(tc_free_memory_rate, "0");
...
 
+CONF_mBool(enable_auto_adjust_pagecache, "true");
+// Memory urget water level, if the memory usage exceeds this level, reduce the size of
+// the Pagecache immediately, it should be between (memory_high_level, 100].
+CONF_mInt64(memory_urgent_level, "85");
+// Memory high water level, if the memory usage exceeds this level, reduce the size of
+// the Pagecache slowly, it should be between [1, memory_urgent_level).
+CONF_mInt64(memory_high_level, "75");
+// Pagecache size adjust period, default 20, it should be between [1, 180].
+CONF_Int64(pagecache_adjuct_period, "20");
```
Comment (chaoyli) [inline be/src/common/config.h:63]: Change the configuration to a mutable variable

Comment (chaoyli) [reply be/src/common/config.h:63]: spell error,`adjust`

File: be/src/storage/olap_server.cpp [hunk 7]
```diff
@@ -176,10 +177,95 @@ Status StorageEngine::start_bg_threads() {
...
+            cache->adjust_capacity(-bytes, kcacheMinSize);
+        }
+    }
+}
+
+void* StorageEngine::_adjust_pagecache_callback(void* arg_this) {
+#ifdef GOOGLE_PROFILER
+    ProfilerRegisterThread();
+#endif
+    GCHelper dec_advisor(config::pagecache_adjuct_period, config::auto_adjust_pagecache_interval_seconds,
```
Comment (chaoyli) [inline be/src/storage/olap_server.cpp:210]: What's the difference between config::pagecache_adjuct_period and config::auto_adjust_pagecache_interval_seconds?

**Comment by zaorangyang** [comment]
run starrocks_be_unittest

**Review by trueeyu — DISMISSED** [review]

**Review by chaoyli — DISMISSED** [review]

**Review by chaoyli — DISMISSED** [review]

**Review by chaoyli — DISMISSED** [review]

**Review by trueeyu — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**